[Flaking Test] UT k8s.io/kubernetes/pkg/controller: storageversiongc

### Which jobs are flaking?

https://prow.k8s.io/job-history/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit
https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-ppc64le/1975012962024296448

### Which tests are flaking?

`k8s.io/kubernetes/pkg/controller/storageversiongc/gc_controller_test.go:Test_StorageVersionDeletedOnLeaseDeletion`

### Since when has it been flaking?

Triage tool shows the earliest failures on [10/06/2025, 07:13:08 AM](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit) as per [Link](https://storage.googleapis.com/k8s-triage/index.html?text=Test_StorageVersionDeletedOnLeaseDeletion). 

### Testgrid link

https://testgrid.k8s.io/sig-release-master-blocking#ci-kubernetes-unit

### Reason for failure (if possible)

```
{Failed  === RUN   Test_StorageVersionDeletedOnLeaseDeletion
    gc_controller_test.go:346: expected IsNotFound error, got: <nil>
--- FAIL: Test_StorageVersionDeletedOnLeaseDeletion (2.00s)
}
```

### Anything else we need to know?

We were able to reproduce this flake locally using stress tool.
**_Steps to reproduce:_** 
1. Install stress tool (# go install golang.org/x/tools/cmd/stress@latest).
2. Clone the k8s repo (git clone https://github.com/kubernetes/kubernetes.git).
3. cd kubernetes
4. go test ./pkg/controller/storageversiongc/ -c -count=1 -race
5. /root/go/bin/stress ./storageversiongc.test -test.run Test_StorageVersionDeletedOnLeaseDeletion

```
[root@cluster-node1 kubernetes]# uname -a
Linux cluster-node1 5.14.0-611.el9.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Aug 22 16:46:08 UTC 2025 x86_64 x86_64 x86_64 GNU/Linux

```
```
[root@cluster-node1 kubernetes]# go test ./pkg/controller/storageversiongc/ -c -count=1 -race
[root@cluster-node1 kubernetes]# /root/go/bin/stress ./storageversiongc.test -test.run Test_StorageVersionDeletedOnLeaseDeletion
5s: 8 runs so far, 0 failures, 8 active
10s: 16 runs so far, 0 failures, 8 active
15s: 32 runs so far, 0 failures, 8 active
20s: 40 runs so far, 0 failures, 8 active
25s: 56 runs so far, 0 failures, 8 active
30s: 64 runs so far, 0 failures, 8 active
35s: 78 runs so far, 0 failures, 8 active
40s: 88 runs so far, 0 failures, 8 active
45s: 96 runs so far, 0 failures, 8 active
50s: 112 runs so far, 0 failures, 8 active
55s: 120 runs so far, 0 failures, 8 active
1m0s: 136 runs so far, 0 failures, 8 active
1m5s: 144 runs so far, 0 failures, 8 active
1m10s: 160 runs so far, 0 failures, 8 active
1m15s: 168 runs so far, 0 failures, 8 active
1m20s: 176 runs so far, 0 failures, 8 active
1m25s: 192 runs so far, 0 failures, 8 active
1m30s: 200 runs so far, 0 failures, 8 active
1m35s: 216 runs so far, 0 failures, 8 active
1m40s: 224 runs so far, 0 failures, 8 active
1m45s: 234 runs so far, 0 failures, 8 active

/tmp/go-stress-20251005T214823-1924337739
I1005 21:50:09.348764 3546701 shared_informer.go:676] "Warning: resync period is too small. Changing it to the minimum allowed value" resyncPeriod="100ms" minimumResyncPeriod="1s"
I1005 21:50:09.349230 3546701 shared_informer.go:676] "Warning: resync period is too small. Changing it to the minimum allowed value" resyncPeriod="100ms" minimumResyncPeriod="1s"
I1005 21:50:09.349476 3546701 gc_controller.go:101] "Starting storage version garbage collector"
--- FAIL: Test_StorageVersionDeletedOnLeaseDeletion (2.00s)
    gc_controller_test.go:346: expected IsNotFound error, got: <nil>
FAIL


ERROR: exit status 1

1m50s: 248 runs so far, 1 failures (0.40%), 8 active
1m55s: 257 runs so far, 1 failures (0.39%), 8 active

/tmp/go-stress-20251005T214823-3448336944
I1005 21:50:19.496947 3547032 shared_informer.go:676] "Warning: resync period is too small. Changing it to the minimum allowed value" resyncPeriod="100ms" minimumResyncPeriod="1s"
I1005 21:50:19.497264 3547032 shared_informer.go:676] "Warning: resync period is too small. Changing it to the minimum allowed value" resyncPeriod="100ms" minimumResyncPeriod="1s"
I1005 21:50:19.497415 3547032 gc_controller.go:101] "Starting storage version garbage collector"
--- FAIL: Test_StorageVersionDeletedOnLeaseDeletion (2.00s)
    gc_controller_test.go:346: expected IsNotFound error, got: <nil>
FAIL


ERROR: exit status 1

2m0s: 272 runs so far, 2 failures (0.74%), 8 active
2m5s: 282 runs so far, 2 failures (0.71%), 8 active
2m10s: 296 runs so far, 2 failures (0.68%), 8 active
```
```
[root@cluster-node1 kubernetes]# cat /tmp/go-stress-20251005T214823-1924337739
I1005 21:50:09.348764 3546701 shared_informer.go:676] "Warning: resync period is too small. Changing it to the minimum allowed value" resyncPeriod="100ms" minimumResyncPeriod="1s"
I1005 21:50:09.349230 3546701 shared_informer.go:676] "Warning: resync period is too small. Changing it to the minimum allowed value" resyncPeriod="100ms" minimumResyncPeriod="1s"
I1005 21:50:09.349476 3546701 gc_controller.go:101] "Starting storage version garbage collector"
--- FAIL: Test_StorageVersionDeletedOnLeaseDeletion (2.00s)
    gc_controller_test.go:346: expected IsNotFound error, got: <nil>
FAIL


ERROR: exit status 1
```

### Relevant SIG(s)

/sig testing

--- a/pkg/controller/storageversiongc/gc_controller_test.go
+++ b/pkg/controller/storageversiongc/gc_controller_test.go
@@ -26,6 +26,7 @@ import (
 	coordinationv1 "k8s.io/api/coordination/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/informers"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/kubernetes/fake"
@@ -328,7 +329,7 @@ func Test_StorageVersionDeletedOnLeaseDeletion(t *testing.T) {
 		},
 	}
 
-	clientset := fake.NewSimpleClientset(lease1, storageVersion)
+	clientset := fake.NewClientset(lease1, storageVersion)
 	_, ctx := ktesting.NewTestContext(t)
 	setupController(ctx, clientset)
 
@@ -337,12 +338,12 @@ func Test_StorageVersionDeletedOnLeaseDeletion(t *testing.T) {
 		t.Fatalf("error deleting lease object: %v", err)
 	}
 
-	// add a delay to ensure controller had a chance to reconcile
-	time.Sleep(2 * time.Second)
-
-	// expect deleted
-	_, err := clientset.InternalV1alpha1().StorageVersions().Get(context.Background(), "k8s.test.resources", metav1.GetOptions{})
-	if !apierrors.IsNotFound(err) {
-		t.Fatalf("expected IsNotFound error, got: %v", err)
+	err := wait.PollUntilContextTimeout(context.Background(), time.Second, 5*time.Second, true, func(ctx context.Context) (bool, error) {
+		// expect deleted
+		_, err := clientset.InternalV1alpha1().StorageVersions().Get(context.Background(), "k8s.test.resources", metav1.GetOptions{})
+		return apierrors.IsNotFound(err), nil
+	})
+	if err != nil {
+		t.Fatalf("expected IsNotFound error, but got error: %v", err)
 	}
 }

